analyzer/src/steps/typing.rs

```diff
@@ -146,67 +155,160 @@ fn ascribe_types(
                 segment: lit.segment.clone(),
             }
         }
+        Expr::TemplateString(tpl) => {
+            if tpl.parts.is_empty() {
+                return TypedExpr {
+                    kind: ExprKind::Literal(LiteralValue::String("".to_owned())),
+                    ty: STRING,
+                    segment: tpl.segment(),
+                };
+            }
+            let plus_method = exploration
+                .engine
+                .get_method_exact(
+                    STRING,
+                    name_operator_method(BinaryOperator::Plus),
+                    &[STRING],
+                    STRING,
+                )
+                .expect("string type should have a concatenation method")
+                .definition;
+            let mut it = tpl.parts.iter().map(|part| {
+                let acc = ascribe_types(
+                    exploration,
+                    relations,
+                    diagnostics,
+                    env,
+                    part,
+                    state.without_local_type(),
+                );
+                convert_into_string(acc, exploration, diagnostics, state)
+            });
+            let acc = it.next().unwrap();
+            it.fold(acc, |acc, current| {
+                let segment = acc.segment.start..current.segment.end;
+                TypedExpr {
+                    kind: ExprKind::MethodCall {
+                        callee: Box::new(acc),
+                        arguments: vec![current],
+                        definition: plus_method,
+                    },
+                    ty: STRING,
+                    segment,
+                }
+            })
+        }
+        Expr::Assign(assign) => {
+            let rhs = ascribe_types(
+                exploration,
+                relations,
+                diagnostics,
+                env,
+                &assign.value,
+                state.with_local_type(),
+            );
+            let symbol = env.get_raw_symbol(assign.segment()).unwrap();
+            let var_obj = exploration
+                .ctx
+                .get(relations, state.source, symbol)
+                .unwrap();
+            let var_ty = var_obj.type_id;
+            let rhs_type = rhs.ty;
+            let rhs = match unify_and_map(
+                rhs,
+                var_ty,
+                &mut exploration.typing,
+                &exploration.engine,
+                state,
+                diagnostics,
+            ) {
+                Ok(rhs) => {
+                    if !var_obj.can_reassign {
```

Comment:
Functions and some other types are not reassignables, so two diagnostics would be raised instead of one (one because a function cannot be reassigned, and the other because a function does not accept the right-hand side expression).